realm/realm-library/src/main/cpp/jni_util/java_method.hpp

```diff
@@ -30,11 +31,28 @@ class JavaMethod {
         : m_method_id(nullptr)
     {
     }
-    JavaMethod(JNIEnv* env, jclass cls, const char* method_name, const char* signature, bool static_method = false);
-    JavaMethod(JNIEnv* env, jobject obj, const char* method_name, const char* signature);
-    JavaMethod(JNIEnv* env, const char* class_name, const char* method_name, const char* signature,
+    JavaMethod(JNIEnv* env, JavaClass const& cls, const char* method_name, const char* signature,
                bool static_method = false);
 
+    // From https://developer.android.com/training/articles/perf-jni.html
+    // The class references, field IDs, and method IDs are guaranteed valid until the class is unloaded. Classes are
+    // only unloaded if all classes associated with a ClassLoader can be garbage collected, which is rare but will not
+    // be impossible in Android. Note however that the jclass is a class reference and must be protected with a call
+    // to NewGlobalRef (see the next section).
+    //
+    // BUT THERE ARE BUGS. See below:
+    //
+    // NOTE: For anyone wants to implement this method, please DON'T. There might be a bug in JVM implementation that
+    // the jmethodID retrieved from jobject's jclass would be invalid under some certain conditions.
+    // See https://github.com/realm/realm-java/issues/4964 for how to reproduce it.
+    // JavaMethod(JNIEnv* env, jobject obj, const char* method_name, const char* signature);
+
+    // For this constructor, no evidence shows that the jmethodID will be invalided if there is no ref to the jclass.
```

Comment:
I think this might be better:
"There is no evidence that jmethodId will be invalid when there is no ref to the jclass.  Just in case, though...
... to make sure we don't encounter another JVM bug."
Just a suggestion.